vmm: Introduce user-defined memory regions

sboeuf opened the pull request:
The goal for this pull request is to introduce a new CLI/API option called `--memory-zone`. This new option works together with the existing `--memory` option, as it gives the user more control over the way the guest memory is backed from a host perspective.
This is an option dedicated for advanced users who want to push the limits of what is possible with Cloud-Hypervisor.

The use of `--memory-zone` will be allowed only when the memory size is 0. Here is a quick example how this option can be used:
```
--memory size=0 --memory-zone size=1G size=4G
```

Fixes #1601

likebreath commented on vmm/src/memory_manager.rs:
```diff
@@ -252,45 +263,149 @@ impl MemoryManager {
+                        ram_region_offset += custom_region_sub_size;
+                        pull_next_custom_region = true;
+
+                        custom_region_sub_size
+                    } else {
+                        custom_region_offset += ram_region_sub_size;
+                        ram_region_consumed = true;
+
+                        ram_region_sub_size
+                    };
+
+                    mem_regions.push(MemoryManager::create_ram_region(
+                        &custom_region.file,
+                        file_offset,
+                        region_start,
+                        region_size,
+                        false,
+                        prefault,
+                        custom_region.shared,
+                        custom_region.hugepages,
+                    )?);
+
+                    if pull_next_custom_region {
+                        // Get the next custom region and reset the offset.
+                        custom_region_offset = 0;
+                        if custom_regions.is_empty() {
+                            exit = true;
+                            break;
+                        }
+                        custom_region = custom_regions.remove(0);
+                    }
+
+                    if ram_region_consumed {
+                        break;
+                    }
+                }
+
+                if exit {
+                    break;
+                }
```
Fitting custom regions into ram regions of various sizes seems to be inevitably convoluted. May be it is worth splitting this piece of code into a separate method, something like `fn create_memory_regions_from_custom_regions(&custom_regions, &ram_regions) -> Vector<memory_regions>`, and adding some comments to explain what it does in general.

sboeuf replied:
Yes, I will create two functions to reduce the complexity of the `MemoryManager::new()`.

likebreath commented:
Nit: typos in the PR discription here. I believe you meant `--memory-region` (not `--memory-config`). 

Great to read code about memory manager. Nice work. Thanks.

likebreath commented on src/main.rs:
```diff
@@ -114,6 +114,18 @@ fn create_app<'a, 'b>(
+            Arg::with_name("memory-region")
+                .long("memory-region")
+                .help(
+                    "Custom memory region parameters \
+                     \"size=<guest_memory_region_size>,mergeable=on|off,\
+                     shared=on|off,hugepages=on|off\"",
```
I am assuming the use of `--memory-region` option will be mostly helpful when users specify explicit backing files. If that's true, do we want to expose the `file` parameter explicitly here?

sboeuf replied:
Oh yes I missed that. I'll fix it.

sboeuf commented on vmm/src/memory_manager.rs:
```diff
@@ -513,13 +518,17 @@ impl MemoryManager {
             Some(ref file) => {
                 let f = if file.is_dir() {
+                    // Override file offset as it does not apply in this case.
+                    file_offset = 0;
```
Yeah or maybe even an `info!()`.

sboeuf commented on vmm/src/memory_manager.rs:
```diff
@@ -513,13 +518,17 @@ impl MemoryManager {
                     unsafe { libc::unlink(path_ptr) };
-                    unsafe { File::from_raw_fd(fd) }
+                    let f = unsafe { File::from_raw_fd(fd) };
+                    f.set_len(size as u64).map_err(Error::SharedFileSetLen)?;
```
I should split this into a dedicated commit. The point is that I think there was a bug as we were truncating the file even when it was a real file (and not a directory). There's no reason to truncate an existing file, as it can be mapped in two (or more) parts (basically what the PR is doing).

sboeuf commented:
@likebreath 
>Nit: typos in the PR discription here. I believe you meant --memory-region (not --memory-config).

Thanks for catching that!

>Great to read code about memory manager. Nice work. Thanks.

Cool! Thanks for the review.

sameo commented on src/main.rs:
```diff
@@ -114,6 +114,18 @@ fn create_app<'a, 'b>(
         )
+        .arg(
+            Arg::with_name("memory-region")
+                .long("memory-region")
```
This is almost nit-picking, but in order to not confuse it with the `vm-memory` crate memory region terminology, could we rename that one? I propose `--memory-zone` instead.

The user should not be confused with the `memory-region` terminology as hopefully this is transparent to users, but from a code perspective we end up talking about `memory_regions`, `ext_regions` and `custom_regions`. While the first ones actually mapped to `vm-memory` regions, the last one does not. It describes something that could be split into several `vm-memory` regions. That something could be called `memory_zones` instead.

sboeuf replied:
Sounds good!

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -246,50 +261,183 @@ impl BusDevice for MemoryManager {
+                        prefault,
+                        config.shared,
+                        config.hugepages,
+                    )?);
```
I think that in order to simplify the code flow, we could internally consider the `--memory size > 0` as a special case of `--memory size=0 --memory-zones size=1G size=4G`.

Essentially, `--memory size=1G` is equivalent to `--memory size=0 --memory-zone size=1G`, i.e we could build a single zone for the `--memory size=1G` case and use the same code path for both `size=0` and `size > 0`.

sboeuf replied:
Yes I agree that'd be nicer. I hope you don't mind me adding this as a new commit instead of trying to rebase through existing commits.

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -309,15 +457,19 @@ impl MemoryManager {
-                    config.hugepages,
-                )?);
+
+                if !use_custom_regions {
```
Are we planning to support a `hotplug_method` and `hotplug_size` per configured memory zone? Or will both be global parameters that apply for all zones?

sboeuf replied:
No I think we talked about using `hotplug_method` and `hotplug_size` from `MemoryConfig` (or `--memory`) as global flags that will apply for all memory zones. Otherwise this would force us to define one more parameter to define global options for the memory zones, and I don't think we want that.

sboeuf commented:
@sameo I've addressed all your comments and I decided to send the entire PR with the snapshot/restore support too. This way, you can see all the simplifications/factorizations I've been able to achieve through these multiple commits.

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -246,51 +255,154 @@ impl BusDevice for MemoryManager {
+    /// same backing file if there's a hole in the address space between two
+    /// RAM ranges.
+    /// One example might be ram_regions containing 2 regions (0-3G and 4G-6G)
+    /// and zones containing two regions (size 1G and size 4G).
```
```suggestion
    /// and zones containing two zones (size 1G and size 4G).
```

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -246,51 +255,154 @@ impl BusDevice for MemoryManager {
+    /// RAM ranges.
+    /// One example might be ram_regions containing 2 regions (0-3G and 4G-6G)
+    /// and zones containing two regions (size 1G and size 4G).
+    /// This function will create 3 resulting regions:
```
```suggestion
    /// This function will create 3 resulting memory regions:
```

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -1137,6 +1137,17 @@ impl MemoryManager {
         &self.sgx_epc_region
     }
+
+    pub fn file_backed_with_hardlink(f: &File) -> bool {
```
```suggestion
    pub fn is_hardlink(f: &File) -> bool {
```

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -1543,22 +1554,44 @@ impl Snapshottable for MemoryManager {
+            let mut backing_file = Some(PathBuf::from(format!("memory-region-{}", index)));
+            if let Some(file_offset) = region.file_offset() {
+                if (region.flags() & libc::MAP_SHARED == libc::MAP_SHARED)
+                    && Self::file_backed_with_hardlink(file_offset.file())
```
Why is it the case only when the file is backed by a hard link?

sboeuf replied:
When the file is backed by at least one hard link, that means the user has a way to retrieve access the file so that a restore will rely on it.
When the file is not backed by any hard link, that means CH created a temporary file with no link, either using the backing directory provided by the user or using `create_memfd()` function directly. In both these cases, there will be no resulting file that can be accessed by the user, which means these files won't be usable on the restore path, therefore we need to create a `memory-region-X` file for storing the content of this memory region.

sameo replied:
> When the file is backed by at least one hard link, that means the user has a way to retrieve access the file so that a restore will rely on it.

That makes sense now, thanks.

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -1543,22 +1554,44 @@ impl Snapshottable for MemoryManager {
         let guest_memory = self.guest_memory.memory();
 
-        let mut memory_regions: Vec<MemoryRegion> = Vec::with_capacity(10);
+        let mut memory_regions: Vec<MemoryRegion> = Vec::new();
```
That seems to be fixing a limitation that was not needed, right?

sboeuf replied:
Yes. Do you want this to be part of a separate commit?

sameo replied:
No that's ok.

sameo commented on vmm/src/memory_manager.rs:
```diff
@@ -1543,22 +1554,44 @@ impl Snapshottable for MemoryManager {
 
+        // Store locally this list of regions as it will be used through the
+        // Transportable::send() implementation.
+        self.snapshot_memory_regions = memory_regions.clone();
```
I understand this saves us from deserializing the Snapshot from `Transportable::send()`, so this should be more explicit in the comment. This is an optimization, not a mandatory setting.

sboeuf replied:
Ok I'll expand the comment!

sameo commented on src/main.rs:
```diff
@@ -114,6 +114,18 @@ fn create_app<'a, 'b>(
                 .group("vm-config"),
         )
+        .arg(
+            Arg::with_name("memory-zone")
```
One last nitpick, sorry for not commenting on that one earlier: Should we use `--memory-zones` instead of `--memory-zone`?

sboeuf replied:
It's really up to you. Based on our current CLI choices, like `--disk`, `--net`, ... we always kept the singular even if we can specify multiple instance of each. Maybe moving everything to plural would make sense at some point, but we could make it as part of a dedicated PR.
Your call :)

sameo reviewed (approved):
One nitpick left, but otherwise LGTM.
